File: python3/_209_minimum_size_subarray_sum.py
#
# @lc app=leetcode id=209 lang=python3
#
# [209] Minimum Size Subarray Sum
#


# @lc code=start
class Solution:
    def minSubArrayLen(self, target: int, nums: list[int]) -> int:
        ans = float("inf")
        l = 0
        r = 0
        s = nums[0]
        while r < len(nums):
            if s < target:
                r += 1
                try:
                    s += nums[r]
                except IndexError:
                    break
            else:
                ans = min(ans, r - l + 1)
                s -= nums[l]
                l += 1
        return ans if ans != float("inf") else 0


# @lc code=end

s = Solution()
assert s.minSubArrayLen(target=7, nums=[2, 3, 1, 2, 4, 3]) == 2
assert s.minSubArrayLen(target=4, nums=[1, 4, 4]) == 1
assert s.minSubArrayLen(target=11, nums=[1, 1, 1, 1, 1, 1, 1, 1]) == 0

File: rust/src/_209_minimum_size_subarray_sum.rs
/*
 * @lc app=leetcode id=209 lang=rust
 *
 * [209] Minimum Size Subarray Sum
 */

// @lc code=start
use std::{
    cmp::{Ordering, min},
    i32,
};

impl Solution {
    pub fn min_sub_array_len(target: i32, nums: Vec<i32>) -> i32 {
        let mut l = 0;
        let mut r = 0;
        let mut sum = nums[0];
        let mut min_len = i32::MAX;
        while r < nums.len() {
            match sum.cmp(&target) {
                Ordering::Less => {
                    r += 1;
                    if r < nums.len() {
                        sum += nums[r];
                    }
                }
                Ordering::Greater | Ordering::Equal => {
                    min_len = min(min_len, (r - l + 1) as i32);
                    sum -= nums[l];
                    l += 1;
                }
            }
        }
        if min_len == i32::MAX { 0 } else { min_len }
    }
}
// @lc code=end

struct Solution;
